CC-31153 Enable conditional writes in connector

subhashiyer9 opened the pull request:
## Problem


## Solution
- Adds a config to enable conditional writes, the default being true
- The conditional write will be enabled only when scheduled interval is set and store kafka key/headers config is disabled to retain the existing delivery semantics
- On a conditional write exception, a FileExistsException is rethrown , which being a retriable exception will reset the offset and poll again accordingly
- In case of FileExistsException, connector will scan the s3 bucket for next available file which is not present in S3 by making consecutive get object requests. If the get api fails with 403 error, it will simply increment to next offset.
- To enable determining the next available file, the equivalent target file name is cached for initial few records, had the record been the first record in the file. The offset will be reset to the one whose equivalent file does not exist in S3
- Use LinkedHashMap to store `commitFiles` instead of existing `HashMap` to ensure files with smaller start offsets (i.e. the ones which were added to the map first) are written to S3 first. This is required to ensure we safely increment to next offset in case of conflict

<!--- Mark x in the box. -->
##### Does this solution apply anywhere else?
- [ ] yes
- [ ] no

##### If yes, where?


## Test Strategy


<!--- Mark x in the box for all that apply. -->
##### Testing done:
- [x] Unit tests
- [ ] Integration tests
- [ ] System tests
- [x] Manual tests
- [ ] Performance tests - TODO
- [ ] Chaos tests - TODO

## Release Plan
<!--- Describe the release plan for this feature. -->
<!-- Are you backporting or merging to master? -->
<!-- If you are reverting or rolling back, is it safe? --> 
The feature will be released as a new minor release on 10.6.x branch

aniketshrimal reviewed (commented):
Thanks @subhashiyer9 for the changes. Have taken a first pass and left some comments. PTAL!

Also could you update the release plan as well in PR description - are we planning to do a minor or major release with this change?

subhashiyer9 commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java:
```diff
@@ -753,12 +777,56 @@ private void commitFile(String encodedPartition) {
+        }
+        log.debug("File {} already exists, checking for next available file", commitFile);
+      } catch (AmazonS3Exception e) {
+        if (e.getStatusCode() == 403) {
```
This is just a simple fallback in case connector does not have permission to Get objects

subhashiyer9 commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java:
```diff
@@ -753,12 +777,56 @@ private void commitFile(String encodedPartition) {
+      String commitFile = offsetToFilenameMap.get(startOffset);
+      try {
+        if (!offsetToFilenameMap.containsKey(startOffset)) {
+          log.info("Start offset {} not present in offsets map. "
```
In my opinion, it would be good to have this in info level, at least for initial few weeks so that we can debug easily in case of any issues

aniketshrimal replied:
Sure, I think we can log it at info level then.

subhashiyer9 commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java:
```diff
@@ -107,6 +107,9 @@ public class S3SinkConnectorConfig extends StorageSinkConnectorConfig {
   public static final String WAN_MODE_CONFIG = "s3.wan.mode";
   private static final boolean WAN_MODE_DEFAULT = false;
 
+  public static final String ENABLE_CONDITIONAL_WRITES_CONFIG = "enable.conditional.writes";
```
No, conditional writes does not require any additional permissions

subhashiyer9 commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java:
```diff
@@ -317,6 +328,12 @@ private void executeState(long now) {
         }
+
+        // TODO: Set appropriate upper bound
+        if (offsetToFilenameMap.size() < MAX_SCAN_LIMIT) {
```
I have created an internal config for this and keeping the default as 100. Even in cases where this limit is not sufficient, it would still be able to eventually determine the next offset in subsequent put cycles, as the offsets will be incremented by 100 in every cycle

subhashiyer9 commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java:
```diff
@@ -769,6 +772,18 @@ public static ConfigDef newConfigDef() {
+          Type.BOOLEAN,
+          ENABLE_CONDITIONAL_WRITES_DEFAULT,
+          Importance.LOW,
+          "Enable conditional writes during multipart upload",
```
Done. Thanks

subhashiyer9 commented on kafka-connect-s3/src/test/java/io/confluent/connect/s3/S3SinkConnectorConfigTest.java:
```diff
@@ -664,5 +668,33 @@ public void testSendDigestConfigOnWrongValue() {
+
+    properties.put(STORE_KAFKA_HEADERS_CONFIG, "false");
+    properties.put(STORE_KAFKA_KEYS_CONFIG, "true");
+    assertFalse(new S3SinkConnectorConfig(properties).shouldEnableConditionalWrites());
```
Done

aniketshrimal reviewed (commented):
LGTM @subhashiyer9! Thanks for addressing the comments. 

One last thing - the build seems to be failing on last commit due to failure of newly added IT. Could you please look into this? Will approve after a green build.

subhashiyer9 commented:
> LGTM @subhashiyer9! Thanks for addressing the comments.
> 
> One last thing - the build seems to be failing on last commit due to failure of newly added IT. Could you please look into this? Will approve after a green build.

The build is now passing

olegmaroseev commented:
Hello. We have a problem with this PR - sometimes in the parquet we get a pass of exactly one offset.
